R/install_torsten.R

```diff
@@ -0,0 +1,197 @@
+#' # install_torsten(cores = 4)
+#'
+#' cpp_options <- list(
+#'   "CXX" = "clang++",
+#'   "CXXFLAGS+= -march=native",
+#'   PRECOMPILED_HEADERS = TRUE
+#' )
+#' # install_torsten(cores = 4, cpp_options = cpp_options)
+#' }
+#'
+
+#'## install_torsten version without unexported dependencies
+
+latest_torsten_release <- function() {
+  require(jsonlite)
+  dest_file <- tempfile(pattern = "releases-", fileext = ".json")
+  download_url <- "https://api.github.com/repos/metrumresearchgroup/Torsten/releases/latest"
+  release_list_downloaded <- utils::download.file(url = download_url,
+                                                  destfile = dest_file,
+                                                  quiet = TRUE)
+  if (release_list_downloaded != 0) {
+    stop("GitHub download of release list failed.", call. = FALSE)
+  }
+  release <- jsonlite::read_json(dest_file)
+  release$tag_name
+}
+
+get_torsten_release_list <- function() {
+  require(jsonlite)
+  dest_file <- tempfile(pattern = "releases-", fileext = ".json")
+  download_url <- "https://api.github.com/repos/metrumresearchgroup/Torsten/releases"
+  release_list_downloaded <- utils::download.file(url = download_url,
+                                                  destfile = dest_file,
+                                                  quiet = TRUE)
+  if (release_list_downloaded != 0) {
+    stop("GitHub download of release list failed.", call. = FALSE)
+  }
+  release <- jsonlite::read_json(dest_file)
+  sapply(release,
+         function(x){
+           x$tag_name
+         })
+}
+
+install_torsten <- function(dir = NULL,
+                            cores = getOption("mc.cores", 2),
+                            quiet = FALSE,
+                            timeout = 1200,
+                            version = NULL,
+                            release_url = NULL,
+                            cpp_options = list(),
+                            check_toolchain = TRUE
+) {
+
+  ## install_torsten is based on cmdstanr::install_cmdstan. Differences include
+  ## * install_torsten always overwrites pre-existing Torsten installation at the
+  ##.  specified cmdstan_path.
+  ## * install_torsten does not try to automatically append to cpp_options for
+  ##.  M1 Macs. Is this needed?
+  ## * install_torsten does not provide support for WSL. Should it?
+
+  require(cmdstanr)
+  if (check_toolchain) {
+    cmdstanr::check_cmdstan_toolchain(fix = FALSE, quiet = quiet)
+  }
+  if (is.null(dir)) {
+    dir <- file.path(dirname(cmdstanr::cmdstan_default_install_path()),
+                     ".torsten")
+  }
+  if (!dir.exists(dir)) {
+    dir.create(dir, recursive = TRUE)
+  }
+  if (!is.null(version)) {
+    if (!is.null(release_url)) {
+      warning("version and release_url shouldn't both be specified!",
+              "\nrelease_url will be ignored.", call. = FALSE)
+    }
+
+    release_list <- get_torsten_release_list()
+    release <- release_list[grep(version, release_list)]
```

Comment:
`version` is user-supplied. They could pass something like "torsten" or even ".", and then this would give multiple values for `release`.

Instead we could 1) check that `version` is a scalar and 2) use `%in%` to verify that it's in the known set of releases.

Reply:
Update after chat with @billgillespie : part of the thinking of using `grep` was that it allows getting around the tag naming inconsistency (i.e. caller doesn't need to pass "torsten_").  So, sticking with `grep` makes sense; it'd just need to signal an error if more than one matching release is found.  Also, this should probably used `fixed = TRUE` so that `version` isn't treated as a regex.

Reply:
Added a test for the case where multiple releases are matched by the `version` argument:
`    if(length(release) > 1){`
`      stop("The version argument, ", version, ", matches multiple Torsten version names", `
`           call. = FALSE)`
`    }`